lib/Sema/CSDiagnostics.cpp

```diff
@@ -2430,6 +2434,97 @@ bool ContextualFailure::tryTypeCoercionFixIt(
   return false;
 }
 
+bool ContextualFailure::tryProtocolConformanceFixIt(
+    InFlightDiagnostic &diagnostic) const {
+  auto innermostTyCtx = getDC()->getInnermostTypeContext();
+  if (!innermostTyCtx)
+    return false;
+
+  auto nominal = innermostTyCtx->getSelfNominalTypeDecl();
+  if (!nominal)
+    return false;
+
+  // We need to get rid of optionals and parens as it's not relevant when
+  // printing the diagnostic and the fix-it.
+  auto unwrappedToType =
+      ToType->lookThroughAllOptionalTypes()->getWithoutParens();
+
+  // If the protocol requires a class & we don't have one (maybe the context
+  // is a struct), then bail out instead of offering a broken fix-it later on.
+  auto requiresClass = false;
+  ExistentialLayout layout;
+  if (unwrappedToType->isExistentialType()) {
+    layout = unwrappedToType->getExistentialLayout();
+    requiresClass = layout.requiresClass();
+  }
+
+  if (requiresClass && !FromType->is<ClassType>()) {
+    return false;
+  }
+
+  // We can only offer a fix-it if we're assigning to a protocol type and
+  // the type we're assigning is the same as the innermost type context.
+  bool shouldOfferFixIt = nominal->getSelfTypeInContext()->isEqual(FromType) &&
+                          unwrappedToType->isExistentialType();
+  if (!shouldOfferFixIt)
+    return false;
+
+  diagnostic.flush();
+
+  // Let's build a list of protocols that the context does not conform to.
+  SmallVector<std::string, 8> missingProtoTypeStrings;
+  for (auto protocol : layout.getProtocols()) {
+    if (!getTypeChecker().conformsToProtocol(
+            FromType, protocol->getDecl(), getDC(),
+            ConformanceCheckFlags::InExpression)) {
+      missingProtoTypeStrings.push_back(protocol->getString());
+    }
+  }
+
+  // If we have a protocol composition type and we don't conform to all
+  // the protocols of the composition, then store the composition directly.
+  // This is because we need to append 'Foo & Bar' instead of 'Foo, Bar' in
+  // order to match the written type.
+  if (auto compositionTy = unwrappedToType->getAs<ProtocolCompositionType>()) {
+    if (compositionTy->getMembers().size() == missingProtoTypeStrings.size()) {
+      missingProtoTypeStrings = {compositionTy->getString()};
+    }
+  }
+
+  assert(!missingProtoTypeStrings.empty() &&
+         "type already conforms to all the protocols?");
+
+  // If we didn't have a protocol composition type, it means we only have a
+  // single protocol, so just use it directly. Otherwise, construct a comma
+  // separated list of missing types.
+  std::string protoString;
+  if (missingProtoTypeStrings.size() == 1) {
```

Comment:
Another nit: We have `interleave` function which is could be used here instead of `if/else`

Reply:
or I can use `llvm::join` directly which will just return the name directly if the size is 1.

Reply:
Or that, sure :)